march_hardware_interface/src/march_hardware_interface.cpp

```diff
@@ -142,33 +146,42 @@ void MarchHardwareInterface::init()
     JointStateHandle jointStateHandle(joint.getName(), &joint_position_[i], &joint_velocity_[i], &joint_effort_[i]);
     joint_state_interface_.registerHandle(jointStateHandle);
 
-    // Create position joint interface
-    JointHandle jointPositionHandle(jointStateHandle, &joint_position_command_[i]);
-
     // Retrieve joint (soft) limits from the urdf
     JointLimits limits;
     getJointLimits(model.getJoint(joint.getName()), limits);
 
-    // Create joint limit interface
-    PositionJointSoftLimitsHandle jointLimitsHandle(jointPositionHandle, limits, soft_limits_[i]);
-    positionJointSoftLimitsInterface.registerHandle(jointLimitsHandle);
+    if (joint.getActuationMode() == march4cpp::ActuationMode::position)
+    {
+      // Create position joint interface
+      JointHandle jointPositionHandle(jointStateHandle, &joint_position_command_[i]);
+      position_joint_interface_.registerHandle(jointPositionHandle);
+
+      // Create joint limit interface
+      PositionJointSoftLimitsHandle jointLimitsHandle(jointPositionHandle, limits, soft_limits_[i]);
+      positionJointSoftLimitsInterface.registerHandle(jointLimitsHandle);
+    }
+    else if (joint.getActuationMode() == march4cpp::ActuationMode::torque)
+    {
+      // Create effort joint interface
+      JointHandle jointEffortHandle(jointStateHandle, &joint_effort_command_[i]);
+      effort_joint_interface_.registerHandle(jointEffortHandle);
 
-    position_joint_interface_.registerHandle(jointPositionHandle);
+      // Create joint effort limit interface
+      EffortJointSoftLimitsHandle jointEffortLimitsHandle(jointEffortHandle, limits, soft_limits_[i]);
+      effortJointSoftLimitsInterface.registerHandle(jointEffortLimitsHandle);
+    }
 
     // Set the first target as the current position
     this->read();
     joint_velocity_[i] = 0;
     joint_effort_[i] = 0;
+    joint_effort_command_[i] = 0;
```

Comment:
Shouldn't the effort command only be set when actuationmode is torque? Same as position command the line below: shouldn't it be set only when actuationmode is position?

Reply:
It doesn't really matter. The effort or position command is only sent towards the IMC when in that certain mode.